### PR TITLE
enh(resource): add notification policy for hosts

### PR DESCRIPTION
**Pull request by adr-mo**
DRAFT of the Host notification policy - Missing unit test for the Use Case

### CODE REVIEW

File: src/Centreon/Domain/Contact/Contact.php [hunk 1]
```diff
@@ -66,6 +66,10 @@ class Contact implements UserInterface, ContactInterface
     public const ROLE_CONFIGURATION_SERVICES_SERVICE_GROUPS_READ_WRITE =
         'ROLE_CONFIGURATION_SERVICES_SERVICE_GROUPS_RW';
     public const ROLE_CONFIGURATION_SERVICES_SERVICE_GROUPS_READ = 'ROLE_CONFIGURATION_SERVICES_SERVICE_GROUPS_R';
+    public const ROLE_CONFIGURATION_USERS_CONTACTS__USERS_READ_WRITE = 'ROLE_CONFIGURATION_USERS_CONTACTS__USERS_RW';
+    public const ROLE_CONFIGURATION_USERS_CONTACTS__USERS_READ = 'ROLE_CONFIGURATION_USERS_CONTACTS__USERS_R';
```
Comment (kduret) [inline src/Centreon/Domain/Contact/Contact.php:70]:
```suggestion
    public const ROLE_CONFIGURATION_CONTACTS_READ_WRITE = 'ROLE_CONFIGURATION_USERS_CONTACTS__USERS_RW';
    public const ROLE_CONFIGURATION_CONTACTS_READ = 'ROLE_CONFIGURATION_USERS_CONTACTS__USERS_R';
```

File: src/Centreon/Domain/Contact/Contact.php [hunk 2]
```diff
@@ -66,6 +66,10 @@ class Contact implements UserInterface, ContactInterface
...
+    public const ROLE_CONFIGURATION_USERS_CONTACT_GROUPS_READ_WRITE = 'ROLE_CONFIGURATION_USERS_CONTACT_GROUPS_RW';
+    public const ROLE_CONFIGURATION_USERS_CONTACT_GROUPS_READ = 'ROLE_CONFIGURATION_USERS_CONTACT_GROUPS_R';
```
Comment (kduret) [inline src/Centreon/Domain/Contact/Contact.php:72]:
```suggestion
    public const ROLE_CONFIGURATION_CONTACT_GROUPS_READ_WRITE = 'ROLE_CONFIGURATION_USERS_CONTACT_GROUPS_RW';
    public const ROLE_CONFIGURATION_CONTACT_GROUPS_READ = 'ROLE_CONFIGURATION_USERS_CONTACT_GROUPS_R';
```

File: src/Centreon/Domain/Engine/EngineConfiguration.php [hunk 3]
```diff
@@ -141,4 +150,22 @@ public function hasIllegalCharacters(string $stringToCheck): bool
...
+        return $this->notificationsEnabledOption;
+    }
+
+    /**
+     * @param int $notificationsEnabledOption
+     * @return self
+     */
+    public function setNotificationsEnabledOption(int $notificationsEnabledOption): self
+    {
+        $this->notificationsEnabledOption = $notificationsEnabledOption;
```
Comment (kduret) [inline src/Centreon/Domain/Engine/EngineConfiguration.php:168]: assert int is among notification enum

File: src/Centreon/Domain/HostConfiguration/Host.php [hunk 4]
```diff
@@ -490,4 +499,22 @@ public function getSeverity(): ?HostSeverity
...
+        return $this->notificationsEnabledOption;
+    }
+
+    /**
+     * @param int $notificationsEnabledOption
+     * @return self
+     */
+    public function setNotificationsEnabledOption(int $notificationsEnabledOption): self
+    {
+        $this->notificationsEnabledOption = $notificationsEnabledOption;
```
Comment (kduret) [inline src/Centreon/Domain/HostConfiguration/Host.php:517]: assert int is among notification enum

File: src/Core/Application/Configuration/Notification/Repository/ReadNotificationRepositoryInterface.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+declare(strict_types=1);
+
+namespace Core\Application\Configuration\Notification\Repository;
+
+use Core\Domain\Configuration\Notification\Model\HostNotification;
+
+interface ReadNotificationRepositoryInterface
+{
+    /**
+     * @param int[] $userIds
```
Comment (kduret) [inline src/Core/Application/Configuration/Notification/Repository/ReadNotificationRepositoryInterface.php:30]: add method description

File: src/Core/Application/Configuration/NotificationPolicy/UseCase/FindHostNotificationPolicy.php [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+        /**
+         * Returns the contacts and contactgroups notified for this Host
+         */
+        [
+            'contact' => $notifiedUserIds,
+            'cg' => $notifiedUserGroupIds,
+        ] = $this->legacyRepository->findHostNotifiedUserIdsAndUserGroupIds($hostId);
+        $users = $this->userRepository->findUsersByIds($notifiedUserIds);
+        $usersNotificationSettings = $this->notificationRepository->findHostNotificationsByUserIds($notifiedUserIds);
```
Comment (kduret) [inline src/Core/Application/Configuration/NotificationPolicy/UseCase/FindHostNotificationPolicy.php:92]:
```suggestion
        $usersNotificationSettings = $this->notificationRepository->findHostNotificationSettingsByUserIds($notifiedUserIds);
```

File: src/Core/Application/Configuration/NotificationPolicy/UseCase/FindHostNotificationPolicy.php [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        if ($engineConfiguration === null) {
+            $this->handleEngineHostConfigurationNotFound($hostId, $presenter);
+            return;
+        }
+        if (
+            $engineConfiguration->getNotificationsEnabledOption() ===
+                EngineConfiguration::NOTIFICATIONS_OPTION_DISABLED
+        ) {
+            $realtimeHost->setIsNotificationEnabled(false);
+        }
```
Comment (kduret) [inline src/Core/Application/Configuration/NotificationPolicy/UseCase/FindHostNotificationPolicy.php:111]: add comment to explain why we get engine information

File: src/Core/Domain/Configuration/Notification/Exception/NotificationException.php [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+class NotificationException extends \InvalidArgumentException
+{
+    /**
+     * @param string $event
+     * @return self
+     */
+    public static function badEvent(string $event): self
+    {
+        return new self(sprintf('Invalid event provided (%s)', $event));
```
Comment (kduret) [inline src/Core/Domain/Configuration/Notification/Exception/NotificationException.php:33]: manage translation

File: src/Core/Domain/Configuration/Notification/Model/HostNotification.php [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+use Core\Domain\Configuration\Notification\Exception\NotificationException;
+use Core\Domain\Configuration\Notification\Model\NotificationInterface;
+use Core\Domain\Configuration\TimePeriod\Model\TimePeriod;
+
+class HostNotification implements NotificationInterface
+{
+    public const EVENT_HOST_RECOVERY = 'RECOVERY',
+                 EVENT_HOST_SCHEDULED_DOWNTIME = 'SCHEDULED_DOWNTIME',
+                 EVENT_HOST_FLAPPING = 'FLAPPING',
+                 EVENT_HOST_NONE = 'NONE',
```
Comment (kduret) [inline src/Core/Domain/Configuration/Notification/Model/HostNotification.php:35]: remove it ?

File: src/Core/Domain/RealTime/Model/Host.php [hunk 10]
```diff
@@ -361,6 +366,25 @@ public function setCommandLine(?string $commandLine): self
...
+    public function isNotificationEnabled(): bool
+    {
+        return $this->isNotificationEnabled;
+    }
+
+    /**
+     * @param bool $isNotificationEnabled
+     * @return self
+     */
+    public function setIsNotificationEnabled(bool $isNotificationEnabled): self
```
Comment (kduret) [inline src/Core/Domain/RealTime/Model/Host.php:381]: setNotificationEnabled ?

File: src/Core/Infrastructure/Configuration/NotificationPolicy/Api/FindHostNotificationPolicyController.php [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+     */
+    public function __invoke(
+        int $hostId,
+        FindHostNotificationPolicy $useCase,
+        FindNotificationPolicyPresenterInterface $presenter
+    ): object {
+        /**
+         * Access denied if no rights given to the configuration for the current user
+         */
+        $this->denyAccessUnlessGrantedForApiConfiguration();
```
Comment (kduret) [inline src/Core/Infrastructure/Configuration/NotificationPolicy/Api/FindHostNotificationPolicyController.php:45]: and api realtime

File: src/Core/Infrastructure/Configuration/NotificationPolicy/Api/FindNotificationPolicyPresenter.php [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        $this->responseStatus = $responseStatus;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getResponseStatus(): ?ResponseStatusInterface
+    {
+        return $this->responseStatus;
+    }
```
Comment (kduret) [inline src/Core/Infrastructure/Configuration/NotificationPolicy/Api/FindNotificationPolicyPresenter.php:115]: not needed ?

**Review by jeremyjaouen — APPROVED** [review]

**Review by a-launois — APPROVED** [review]

**Pull request merged**